### PR TITLE
Portable apps specification

### PR DESCRIPTION
**Pull request by denelon**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [x] Are you working against an Issue?

-----

Related to:

*  #182

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2012)

### CODE REVIEW

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:38]:
```suggestion
The default path for installing these packages is "%LOCALAPPDATA%/WinGet_Packages/User" for user based installs. The default path for installing these packages is "%PROGRAMDATA%/WinGet_Packages/Machine" for machine wide installs.
```
The localappdata for a specific user may not be accessible by all users on a machine, so I would recommend these be put into a central location. Perhaps they could all be put into programdata and the registry entries determine whether it is installed for all users or a single user?

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:51]: We may need to specify that the entry here should be created in HKCU for user scope and HKLM for machine scope.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:56]: I think that it would be useful to have the Product Code also be tied to a specific installer/version like it is with MSI installers. Perhaps it could be to the SHA256, or the "PackageVersion"?

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:58]: Could this be an option in user settings? Maybe under future considerations

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:68]: How does this interact with channels? I know it may be outside the scope of the spec, but I think it's important to consider. A user may want both a stable and beta version of a portable app at the same time

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:74]: Edge case - if `winget uninstall {PackageIdentifier} --purge` is set as the uninstall string, and the user no longer has winget installed, what happens?

Would it be beneficial to write the uninstall string as a string of commands to be passed directly to cmd.exe? 

(Probably ignore this comment, it is likely a very rare edge case)

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:80]: Several applications that support portable installs and static installs support multiple commands. I think that they should support more than one command and that the first command in the array should define the default value.

What about apps with monikers? Using vscode as an example - if it were portable type, then the command is `code`, so the rename would be `code.exe` and not `vscode.exe`? Seems a bit confusing.

Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:80]: As mentioned above, we may also want to consider how channels may impact the requirements of renaming files

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:175]: I would suggest that this issue be mentioned, as any path output in the case of remaining files may run into similar feedback -
* https://github.com/microsoft/winget-cli/issues/1977

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:78]: Should "scope" be a valid manifest field when installer type is portable? I would imagine that portable applications should default to user scope and only use machine scope if specified using --scope

Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:78]: Should there be only one `AppsAndFeaturesEntries` for portable packages? I can't imagine a scenario where there should be 2 of them

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:64]: > The first step the Windows Package Manager will perform is an uninstall of the previous version of the package

Manifest Validation may need to warn that `UpgradeBehavior`  should be null or "uninstallPrevious" in the case of a portable installer. It may be confusing otherwise, even though the behavior will be the same.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:38]: @ryfu-msft thoughts?

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:56]: That's a good point.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:58]: 👍

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:68]: Future Considerstions.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:74]: Yeah, I'd say that's an edge case. Updated "Potential Issues".

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:175]: added

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:78]: @ryfu-msft

Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:78]: I think we discussed this before with various scenarios. I don't remember if we landed on this as a semantic requirement or not. 

@ryfu-msft thoughts?

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:64]: @ryfu-msft should we add this as another semantic validation?

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:38]: I think that Program Files is the correct location for machine wide installs, but that is more my intuition than having looked at the current guidance.  I agree that the specific user's local app data is not the right location.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:40]:
```suggestion
The corresponding settings are "PortablePackageUserRoot" and "PortablePackageMachineRoot".
```

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:44]: Do we want to specify how a conflict between identifiers from different sources would be handled?  My first thought is that by default we wouldn't include anything about the source, but if a conflict was detected, we would prepend the source name to the second (and third, etc.) directories.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:51]: And even exactly which architecture we should use when writing it (should an x86 package on an x64 device go into the `SysWow...` path).

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:56]: Product Codes that change over time are actually quite annoying.  I would prefer that we don't do that.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:64]: I think we should not define that we will "uninstall" first; that should be left to the behavior defined in the manifest.  We should attempt to determine if files are in use though before proceeding (not that it can be perfect) and then replace any changed files.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:68]: Channels are meant for use by packages that cannot be installed SxS between stable/beta.  For packages that can be, the thought was to use different package ids.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:74]: I disagree on the default here; we should not purge.  I'm not unmovable, but generally it seems like if we already inform them of the extra files, they can act.

It also makes installing to a non-empty directory very dangerous for the future.  I think generally maybe purge shouldn't work if a custom, existing or non-empty location was chosen to prevent accidental nukage.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:76]: We should probably track whether we created the directory and decide based on that too.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:78]: Yes, scope will come from the user in these cases, so it will probably be ignored in the manifest.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:80]: We could allow the command(s) to define the App Path entrie(s), while not renaming the file.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:112]: "options" being actions that they could take, not "interactive command line questionnaire".

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:44]: I think it's an edge case, but what if a user changes/removes that source? We thought about including the source in one iteration internally but decided against it.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:74]: The concern is if we are calling the uninstall from Apps & Features, the user wouldn't get informed. If we had a way to inform the user, then I'd agree. I just don't know what's possible here.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:112]: Agree, I'm trying to ensure we avoid any user input. My intent with "given options" was just the additional text in the output with the hint for "--force".

Comment (denelon) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:112]:
```suggestion
If the portable application is running when the user performs an upgrade, the user will be informed.
```

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:38]: If we go with Program Files, then Architecture should also be respected to place them in `Program Files` of `Program Files (x86)` respectively

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:56]: > Product Codes that change over time are actually quite annoying. I would prefer that we don't do that.

As long as there aren't any potential issues with matching a program to an ARP entry, or disambiguation, then I would think that having static product codes would be okay. I would assume that the DisplayName and DisplayVersion would always be set, so it may not be an issue unless it conflicts with an already installed application making the package ambiguous

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (ryfu-msft) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:78]: That makes sense to me. We can add an additional manifest validation rule that only applies to the "portable" installerType that verifies only one entry for ```AppsAndFeaturesEntries``` can be specified.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (jedieaston) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:38]: If a portable app expects to be able to write to the directory it is installed in (as many do, since they expect that they are installed per user), putting them in Program Files would probably break stuff since non adminstrators can't write there. 

This may open up a security can of worms, but what if when you installed a portable app for the system winget just put it in everyone's %appdata%? You'd be making several copies of it, but it would improve compatibility. (I'm thinking how the Teams Machine-Wide Installer works.) Alternatively, maybe symlinking the individual files the portable app comes with from Program Files to somewhere the user can write to (so they get the application itself from program files, but any new preferences or other data would be written to a folder they had access to).

Comment (JohnMcPMS) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:38]: I feel that the expectation with a system installation is that all users (even those that have not logged in yet) will have access.  That makes it hard to do anything that involves updating every user's files (as well as being questionable in terms of respecting other users' ownership).  I think that it is also a security problem to allow any user to write to the shared location.

I don't think that we are trying to make every package work in every scenario; and this seems like an area where if the portable needs to write to its current location, it might just not work for system install and non-admin use cases.

**Comment by denelon** [comment]
I believe we've aligned the specification to address the comments that have been made. We're going to have the engineering team do one more review before merging the specification in.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (Trenly) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:79]:
```suggestion
A new Windows Apps & Features entry is created to correctly report the upgraded version for future potential upgrades.
```
Is this a new entry, or is the existing entry from the existing package updated?

Comment (denelon) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:79]: If we uninstall, then the entry is removed. If we can simply upgrade, the existing entry should get updated.

File: doc/specs/#182 - Support for installation of portable standalone apps.md
Comment (JohnMcPMS) [inline doc/specs/#182 - Support for installation of portable standalone apps.md:70]: I don't think that an additional subdirectory should be created; it makes it harder to visually scan the packages directory to find things.  Something more like `PackageIdentifier_Source2` would make it easier to find both the first and second package directories together.

Comment (JohnMcPMS) [reply doc/specs/#182 - Support for installation of portable standalone apps.md:70]: Additionally, I'm not sure that we should do "last writer wins (LWW)" semantics.  I think that is appropriate if we did nothing else, but I would also like to have a command like `winget app-path apply PackageFoo`.  If we had that command, I would think it better to have "first writer wins (FWW)" and allow the user to use the new command to apply the appropriate app-path.

We could start with FWW and an additional parameter to specify LWW for now, then add the app path update command in the future.

Comment (ryfu-msft) [reply doc/specs/#182 - Support for installation of portable standalone apps.md]: I updated the description to reflect your suggestion. I also agree that not having a separate directory just for source would make it easier to find both packages in the same directory. 

I included further clarification for having FWW be our default behavior as well as having an argument to support allowing users to have a LWW behavior if they want that using the --lww argument

**Review by ryfu-msft — APPROVED** [review]

**Pull request merged**
